api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -165,35 +169,57 @@ public void close() {
     }
   }
 
-  private void scheduleTokensRenewal(Tokens currentTokens) {
-    Instant now = Instant.now();
+  private void maybeScheduleTokensRenewal(Tokens currentTokens) {
+    Instant now = clock.get();
+    if (idle(now)) {
+      sleeping.set(true);
+      LOGGER.debug("Sleeping...");
+    } else {
+      scheduleOrExecuteTokensRenewal(currentTokens, now, MIN_REFRESH_DELAY);
+    }
+  }
+
+  private void scheduleOrExecuteTokensRenewal(
+      Tokens currentTokens, Instant now, Duration minRefreshDelay) {
     Instant accessExpirationTime =
         tokenExpirationTime(now, currentTokens.getAccessToken(), defaultAccessTokenLifespan);
     Instant refreshExpirationTime =
         tokenExpirationTime(now, currentTokens.getRefreshToken(), defaultRefreshTokenLifespan);
     Duration delay =
-        nextDelay(now, accessExpirationTime, refreshExpirationTime, refreshSafetyWindow);
-    LOGGER.debug("Scheduling token refresh in {}", delay);
-    tokenRefreshFuture =
-        executor.schedule(
-            () -> {
-              currentTokensStage = renewTokens();
-              currentTokensStage.thenAccept(this::scheduleTokensRenewal);
-            },
-            delay.toMillis(),
-            TimeUnit.MILLISECONDS);
+        nextDelay(
+            now, accessExpirationTime, refreshExpirationTime, refreshSafetyWindow, minRefreshDelay);
+    if (delay.compareTo(MIN_REFRESH_DELAY) < 0) {
+      LOGGER.debug("Refreshing tokens immediately");
+      renewTokens();
+    } else {
+      LOGGER.debug("Scheduling token refresh in {}", delay);
+      tokenRefreshFuture =
+          executor.schedule(this::renewTokens, delay.toMillis(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void renewTokens() {
+    CompletionStage<Tokens> oldTokensStage = currentTokensStage;
+    currentTokensStage =
+        oldTokensStage
+            // try refreshing the current tokens
+            .thenApply(this::refreshTokens)
+            // if that fails, try fetching brand-new tokens
+            .exceptionally(error -> fetchNewTokens())
+            .whenComplete(this::log);
+    currentTokensStage.thenAccept(this::maybeScheduleTokensRenewal);
   }
 
-  private CompletionStage<Tokens> renewTokens() {
-    return currentTokensStage
-        // try refreshing the current tokens
-        .thenApply(this::refreshTokens)
-        // if that fails, try fetching brand-new tokens
-        .exceptionally(error -> fetchNewTokens())
-        .whenComplete(this::log);
+  /** Visible for testing. */
+  boolean idle(Instant now) {
+    return Duration.between(lastAccess, now).compareTo(idleInterval) > 0;
+  }
+
+  private void wakeUp(Instant now) {
+    scheduleOrExecuteTokensRenewal(getCurrentTokens(), now, Duration.ZERO);
```

Comment:
Can you inline this to the (only) call site?

Reply:
I thought having a method would better self-document the intent, but sure. We can also remove `shouldSleep` in fact.